File: app/src/main/java/net/n4dev/treespot/core/AbstractQuery.java
package net.n4dev.treespot.core;

import androidx.annotation.NonNull;

import net.n4dev.treespot.core.api.IEntity;
import net.n4dev.treespot.core.api.IQuery;
import net.n4dev.treespot.db.TreeSpotObjectBox;

import io.objectbox.Box;
import io.objectbox.Property;
import io.objectbox.query.Query;
import io.objectbox.query.QueryCondition;

public abstract class AbstractQuery<T extends IEntity> implements IQuery<T> {

    private final Box<T> paramBox;

    public abstract QueryCondition<T> buildConditions(@NonNull Property<T>[] fields);

    public AbstractQuery(Class<T> klass) {
        paramBox = TreeSpotObjectBox.INSTANCE.getBox(klass);
    }


    @Override
    public QueryCondition<T> buildQuery() {
        return buildConditions(paramBox.getEntityInfo().getAllProperties());
    }
}

File: app/src/main/java/net/n4dev/treespot/core/AbstractQuery.kt
package net.n4dev.treespot.core

import io.objectbox.Box
import io.objectbox.Property
import io.objectbox.query.QueryCondition
import net.n4dev.treespot.core.api.IEntity
import net.n4dev.treespot.core.api.IQuery
import net.n4dev.treespot.db.TreeSpotObjectBox.getBox

/**
 * Base class for querying data based on an IEntity.
 */
abstract class AbstractQuery<T : IEntity>(klass: Class<T>) : IQuery<T> {

    private val paramBox: Box<T> = getBox(klass)
    abstract fun buildConditions(fields: Array<Property<T>>): QueryCondition<T>
    override fun buildQuery(): QueryCondition<T> {
        return buildConditions(paramBox.entityInfo.allProperties)
    }

}